File: apps/android/core/src/main/kotlin/com/discountscreener/core/engine/PricingHistoryMerge.kt
package com.discountscreener.core.engine

import com.discountscreener.core.model.ChartRange
import com.discountscreener.core.model.PricingCandle
import java.time.Instant
import java.time.ZoneOffset
import java.time.temporal.WeekFields

object PricingHistoryMerge {
    fun merge(
        existing: List<PricingCandle>,
        incoming: List<PricingCandle>,
    ): List<PricingCandle> {
        val candlesByKey = LinkedHashMap<PricingCandleKey, PricingCandle>()
        existing.forEach { candle ->
            candlesByKey[candle.key()] = candle
        }
        incoming.forEach { candle ->
            candlesByKey[candle.key()] = candle
        }
        return candlesByKey.values.sortedWith(pricingCandleOrder)
    }

    private val pricingCandleOrder = compareBy<PricingCandle>(
        { it.symbol },
        { it.range },
        { it.candle.epochSeconds },
    )

    private fun PricingCandle.key(): PricingCandleKey = PricingCandleKey(
        symbol = symbol,
        range = range,
        period = semanticPeriod(),
    )

    private fun PricingCandle.semanticPeriod(): String {
        if (range == ChartRange.Day || range == ChartRange.Week) {
            return "epoch:${candle.epochSeconds}"
        }

        val date = Instant.ofEpochSecond(candle.epochSeconds).atZone(ZoneOffset.UTC).toLocalDate()
        return when (range) {
            ChartRange.Month -> "day:$date"
            ChartRange.Year -> {
                val iso = WeekFields.ISO
                "week:${date.get(iso.weekBasedYear())}-${date.get(iso.weekOfWeekBasedYear())}"
            }
            ChartRange.FiveYears,
            ChartRange.TenYears -> "month:${date.year}-${date.monthValue}"
            ChartRange.Day,
            ChartRange.Week -> error("intraday ranges use exact epoch identity")
        }
    }

    private data class PricingCandleKey(
        val symbol: String,
        val range: ChartRange,
        val period: String,
    )
}

File: apps/flutter/packages/ds_core/lib/src/pricing_history_merge.dart
import 'chart_projection.dart';
import 'models.dart';

/// Dedupes pricing candles by semantic period (Android `PricingHistoryMerge`).
class PricingHistoryMerge {
  static List<PricingCandle> merge(
    List<PricingCandle> existing,
    List<PricingCandle> incoming,
  ) {
    final candlesByKey = <String, PricingCandle>{};
    for (final candle in existing) {
      candlesByKey[_key(candle)] = candle;
    }
    for (final candle in incoming) {
      candlesByKey[_key(candle)] = candle;
    }
    final values = candlesByKey.values.toList()
      ..sort((a, b) {
        final s = a.symbol.compareTo(b.symbol);
        if (s != 0) return s;
        final r = a.range.index.compareTo(b.range.index);
        if (r != 0) return r;
        return a.candle.epochSeconds.compareTo(b.candle.epochSeconds);
      });
    return values;
  }

  static String _key(PricingCandle candle) =>
      '${candle.symbol}|${candle.range.name}|${_semanticPeriod(candle)}';

  static String _semanticPeriod(PricingCandle candle) {
    if (candle.range == ChartRange.day || candle.range == ChartRange.week) {
      return 'epoch:${candle.candle.epochSeconds}';
    }
    final date = DateTime.fromMillisecondsSinceEpoch(
      candle.candle.epochSeconds * 1000,
      isUtc: true,
    );
    switch (candle.range) {
      case ChartRange.month:
        return 'day:${date.toIso8601String().substring(0, 10)}';
      case ChartRange.year:
        // ISO week: Monday-based week of year.
        final thursday = date.add(Duration(days: 3 - ((date.weekday + 6) % 7)));
        final firstThursday = DateTime.utc(thursday.year, 1, 4);
        final week1Monday = firstThursday.subtract(
          Duration(days: (firstThursday.weekday + 6) % 7),
        );
        final week =
            1 + thursday.difference(week1Monday).inDays ~/ 7;
        return 'week:${thursday.year}-$week';
      case ChartRange.fiveYears:
      case ChartRange.tenYears:
        return 'month:${date.year}-${date.month}';
      case ChartRange.day:
      case ChartRange.week:
        return 'epoch:${candle.candle.epochSeconds}';
    }
  }
}
